liquidity_pool/src/lib.rs

```diff
@@ -1,13 +1,11 @@
 #![no_std]
 
 mod test;
-pub mod testutils;
 mod token;
 
 use num_integer::Roots;
-use soroban_sdk::{contractimpl, Bytes, BytesN, ConversionError, Env, RawVal, TryFromVal};
+use soroban_sdk::{contractimpl, Address, Bytes, BytesN, ConversionError, Env, RawVal, TryFromVal};
 use token::create_contract;
-use token::{Identifier, Signature};
 
 #[derive(Clone, Copy)]
 #[repr(u32)]
```

Comment:
I think that's a good opportunity to convert this to struct - the data access pattern here is inefficient (6x ledger keys and 6x storage reads/writes)), plus the readability is worse as the values aren't explicitly typed. But we can do this as a followup for the next release as well if that seems to take too long - up to you.

Reply:
Is it actually more efficient if all 6 are in a single struct? I can see putting the tokens into a struct and the reserves into a struct, but all 6 may not work well because the access patterns are different (for example, swap the pool share token contract or the total shares.)

Reply:
Well, from the ledger space standpoint merging everything together is always better. 

We probably could store RW and RO data separately. However, if we think about it, even if we had parallel execution, we'd still need to perform all the operations sequentially (as reserves are affected by every operation), so there is no benefit from this standpoint. 

Now from the fee standpoint I'm pretty sure that we always have one RW entry (reserves), and hence no matter what the pricing is RW+RO would always be more expensive than just a single RW.  We also seemingly need to read all the values every time, so there isn't benefit in terms of data size. So unless I'm missing something about the implementation (which I haven't read too thoroughly) there doesn't seem to be any point in doing anything more granular than a single struct.

Again, fee model isn't there yet, but just from some basic properties it should have (like charging more for more ledger keys, charging more for more ledger space etc) it seems like it's always a good idea to batch the data together as much as possible. The only argument against struct **currently** is that we store the field names. But that should be changed soon.

Reply:
> We also seemingly need to read all the values every time

This isn't true as I mentioned. Swap doesn't touch `TokenShare` or `TotalShares`. But I'm not sure if that's worth the tradeoff of separate data keys and values.

I agree the data model can be improved, but I would prefer to make any changes in a subsequent PR.

Reply:
A separate PR or just an issue are perfectly fine too. 

I see, so `swap` doesn't use these two fields. Maybe they could be put into a separate struct. But I'm still not sure if saving 48 bytes on read in one place is worth permanently adding 32+ bytes of ledger space (for the key) and 32+ bytes read and +1 entry for non-swap operations. In any case, at least condensing these into two structs should be beneficial.

Another consideration is archival - not sure where we land with this, but it feels like bundling correlated data makes things simpler for recovery.

Reply:
https://github.com/stellar/soroban-examples/issues/207